src/Compression/CompressionCodecDeflateQpl.cpp

```diff
@@ -16,59 +17,72 @@ namespace ErrorCodes
     extern const int CANNOT_DECOMPRESS;
 }
 
-std::array<qpl_job *, DeflateQplJobHWPool::MAX_HW_JOB_NUMBER> DeflateQplJobHWPool::hw_job_ptr_pool;
-std::array<std::atomic_bool, DeflateQplJobHWPool::MAX_HW_JOB_NUMBER> DeflateQplJobHWPool::hw_job_ptr_locks;
-bool DeflateQplJobHWPool::job_pool_ready = false;
-std::unique_ptr<uint8_t[]> DeflateQplJobHWPool::hw_jobs_buffer;
-
 DeflateQplJobHWPool & DeflateQplJobHWPool::instance()
 {
     static DeflateQplJobHWPool pool;
     return pool;
 }
 
 DeflateQplJobHWPool::DeflateQplJobHWPool()
-    : random_engine(std::random_device()())
-    , distribution(0, MAX_HW_JOB_NUMBER - 1)
+    : hw_jobs_max_number(0)
+    , random_engine(std::random_device()())
 {
     Poco::Logger * log = &Poco::Logger::get("DeflateQplJobHWPool");
-    UInt32 job_size = 0;
     const char * qpl_version = qpl_get_library_version();
 
+    // loop all configured workqueue size to get maximum job number.
+    accfg_ctx *ctx_ptr = nullptr;
+    auto ctx_status = accfg_new(&ctx_ptr);
+    if (ctx_status == 0)
+    {
+        auto *dev_ptr = accfg_device_get_first(ctx_ptr);
+        while (nullptr != dev_ptr) { // loop all devices
+            for (auto *wq_ptr = accfg_wq_get_first(dev_ptr); \
+                wq_ptr != nullptr; \
+                wq_ptr = accfg_wq_get_next(wq_ptr))
+                hw_jobs_max_number += accfg_wq_get_size(wq_ptr);
+            dev_ptr = accfg_device_get_next(dev_ptr);
+        }
+    }
+
+    if (hw_jobs_max_number == 0)
```

Comment:
I think a distinction between
- `ctx_status != 0` (l. 36, which seems to indicate a configuration issue or the like) and
- `ctx_status == 0` (l. 36) yet `hw_jobs_max_number == 0` (which seems to indicate correct configuration in general but a lack of hardware resources)

would be helpful. Maybe we can have different logging messages (l. 51) for that?

Reply:
Done